### PR TITLE
[flang][OpenMP] Support tasks' implicit firstprivate DSA

### LINKED ISSUE #64480: [Flang][OpenMP]Variable used in TASK construct is not implicit firstprivate
```
Version of flang-new : 17.0.0(359f170f5f712ee714193b46bad45a45656b2c59)
```

Variable used in `TASK` construct is not implicit `FIRSTPRIVATE`.

According to the following specification, the variable(a) in the test program is `FIRSTPRIVATE`.  
- OpenMP Application Programming Interface Version 4.5 : subclause 2.15.1.1

The following are the test program, results of Flang-new, Gfortran and ifort compilation and execution.

omp3_task_024_2.f90:
```fortran
program main
!$omp parallel
  call sub()
!$omp end parallel
  print *,"pass"
end program main

subroutine sub
  integer::a
  a=5
!$omp task
  write(6,*) "1  a = ", a
  if (a.eq.5) then
     a=a+10
  else
     print *,"NG-1:a=",a
  end if
  write(6,*) "2  a = ", a
!$omp end task
!$omp taskwait
  write(6,*) "3  a = ", a
  if (a.ne.5) then
     print *,"NG-2:a=",a
  end if
end subroutine sub
```

```
$ export OMP_NUM_THREADS=2; flang-new -fopenmp omp3_task_024_2.f90; ./a.out
 1  a =  5
 2  a =  15
 3  a =  15
 NG-2:a= 15
 1  a =  5
 2  a =  15
 3  a =  15
 NG-2:a= 15
 pass
$
```

```
$ export OMP_NUM_THREADS=2; gfortran -fopenmp omp3_task_024_2.f90; ./a.out
 1  a =            5
 2  a =           15
 3  a =            5
 1  a =            5
 2  a =           15
 3  a =            5
 pass
$
```

```
$ export OMP_NUM_THREADS=2; ifort -qopenmp omp3_task_024_2.f90; ./a.out
 1  a =            5
 2  a =           15
 3  a =            5
 1  a =            5
 2  a =           15
 3  a =            5
 pass
$
```

### PR DESCRIPTION
**Pull request by luporl**
Handle implicit firstprivate DSAs on task generating constructs.

Fixes https://github.com/llvm/llvm-project/issues/64480

### CODE REVIEW

**Comment by kiranchandramohan** [comment]
I am away this week, will come back to this next week.

File: flang/test/Lower/OpenMP/FIR/default-clause.f90 [hunk 1]
```diff
@@ -213,11 +213,10 @@ subroutine nested_default_clause_tests
 !CHECK: omp.terminator
 !CHECK: }
 !CHECK: omp.parallel {
-!CHECK: %[[PRIVATE_INNER_Z:.*]] = fir.alloca i32 {bindc_name = "z", pinned, uniq_name = "_QFnested_default_clause_testsEz"}
 !CHECK: %[[PRIVATE_INNER_W:.*]] = fir.alloca i32 {bindc_name = "w", pinned, uniq_name = "_QFnested_default_clause_testsEw"} 
 !CHECK: %[[PRIVATE_INNER_X:.*]] = fir.alloca i32 {bindc_name = "x", pinned, uniq_name = "_QFnested_default_clause_testsEx"}
 !CHECK: %[[temp_1:.*]] = fir.load %[[PRIVATE_INNER_X]] : !fir.ref<i32>
-!CHECK: %[[temp_2:.*]] = fir.load %[[PRIVATE_INNER_Z]] : !fir.ref<i32>
+!CHECK: %[[temp_2:.*]] = fir.load %[[PRIVATE_Z]] : !fir.ref<i32>
```
Comment (NimishMishra) [inline flang/test/Lower/OpenMP/FIR/default-clause.f90:219]: Thanks for this. Indeed `shared(z)` needs to have `PRIVATE_Z` and not `PRIVATE_INNER_Z`.

**Review by NimishMishra — APPROVED** [review]
Thanks a lot for these changes. I went through the changes; the handling of symbols in semantics seems particularly fine.

LGTM!  The windows builtbot is failing, though Linux is passing.

**Comment by NimishMishra** [comment]
Also, the PR shows "[luporl] wants to merge 1 commit into [**llvm:users/luporl/refactor-default-test**]from [luporl:luporl-omp-implicit]"

Is this merge not in `main`?

**Comment by luporl** [comment]
@NimishMishra Thanks for the review!

It seems the buildbot errors on Windows are all of this type: `fatal error C1002: compiler is out of heap space in pass 2`.

> Is this merge not in main?

It is not. It depends on https://github.com/llvm/llvm-project/pull/85978, that depends on https://github.com/llvm/llvm-project/pull/72510. As tasks' implicit firstprivate happens only when no default clause is specified, I thought it would be better to start from the default clause fix, in https://github.com/llvm/llvm-project/pull/72510.

Actually, https://github.com/llvm/llvm-project/pull/78283 is also needed to avoid issues with threadprivate.

File: flang/lib/Semantics/resolve-directives.cpp [hunk 2]
```diff
@@ -2012,34 +2012,87 @@ void OmpAttributeVisitor::Post(const parser::Name &name) {
...
       }
-      if (hasDataSharingAttr) {
-        if (defaultDSASymbols.size())
-          symbol = &MakeAssocSymbol(symbol->name(), *defaultDSASymbols.back(),
+
+      // When handling each implicit rule, either a new private symbol is
+      // declared or the last declared symbol is used.
+      // In the latter case, it's necessary to insert a new symbol in the scope
+      // being processed, associated with the last declared symbol, to avoid
+      // "inheriting" the enclosing context's symbol and its flags.
```
Comment (kiranchandramohan) [inline flang/lib/Semantics/resolve-directives.cpp:2036]: Can you expand the `latter case`? On a first look it feels like we should be OK with just inheriting the enclosing context's symbol and its flags. Probably you have to say something like we have to create a new symbol to capture the fact that although we are using the last declared symbol its datasharing attribute could be different in this scope and give an example.

**Review by kiranchandramohan — APPROVED** [review]
LG. Nice work.

File: flang/lib/Semantics/resolve-directives.cpp (hunk 2 above)
Comment (luporl) [reply flang/lib/Semantics/resolve-directives.cpp]: I have expanded the explanation of why a new symbol is needed when inheriting the enclosing context's symbol.
Please check if it is clear now.

Comment (kiranchandramohan) [reply flang/lib/Semantics/resolve-directives.cpp]: Thanks @luporl. It is very clear now.

**Comment by luporl** [comment]
Windows' CI error seems unrelated:
`llvm\lib\Target\AMDGPU\SIModeRegisterDefaults.cpp(239): error C2131: expression did not evaluate to a constant`

**Pull request merged**